### PR TITLE
JS: Improve stability of JS model building (prevent cycles)

### PR DESCRIPTION
**Pull request by matthiasblaesing**
As analyzed by @akronenw certain JS files can lead the JS model generator of NetBeans
to create a cyclic structure. The structure looks like this:

```javascript
var Base = new function() {
     // a lot of code

    function Base() {
        // some code
    }

    // more code
}
```

This is turn is at least partially responsible for seemingly endless scanning.

The implemented fix added an assert that detects cycles in the generated model and fixes the
one detected when the `paper-full.js` file is parsed.

In addition a wrong assertion in the GraalJS fork is disabled, that triggered on angular projects
and prevented usage of NetBeans with activated assertions there.

This is not a full fix, but should improve the situation and make it easier to find other
problems in that area by just using NetBeans with assertions enabled (for example by running
a nightly build).

Closes: #4745

### CODE REVIEW

**Comment by matthiasblaesing** [comment]
This is build on top of #5765, which has to be merged first.

File: webcommon/javascript2.model/src/org/netbeans/modules/javascript2/model/JsObjectImpl.java [hunk 1]
```diff
@@ -192,7 +191,17 @@ public JsObject getParent() {
...
+    @SuppressWarnings({"NestedAssignment", "AssertWithSideEffects"})
     public void setParent(JsObject newParent) {
+        boolean assertionsEnabled = false;
+        assert assertionsEnabled = true;
+        if (assertionsEnabled) {
+            for(JsObject checkParent = newParent; checkParent != null; checkParent = checkParent.getParent()) {
+                if(checkParent == this) {
+                    throw new AssertionError("Cycle detected in " + getFileObject().getPath());
+                }
+            }
```
Comment (junichi11) [inline webcommon/javascript2.model/src/org/netbeans/modules/javascript2/model/JsObjectImpl.java:203]: Nitpick: formatting

**Review by junichi11 — APPROVED** [review]
Great work!
I've verified that `Base()` is shown in the Navigator.
Before:
nb-gh-5786-before
After:
nb-gh-5786-after
Maybe, we can also add a unit test to `JsStructureScannerTest` of `javascript2.editor` although I'm not sure about the JS area.

**Comment by matthiasblaesing** [comment]
@junichi11 thank you for the review. You are right in both points and I updated this PR in place. The formatting as updated using NetBeans formatter and I added a test, that tests the detected structure. This adds a second perspective to the one already checked via `ModelTest#testObjectNameMatchingNestedFunction`, which might be helpful later, when updates are necessary.

**Comment by junichi11** [comment]
Perfect!

**Comment by matthiasblaesing** [comment]
@junichi11 thank you for review,  @akronenw thank you for testing. I'll merge with all test runs being green.

**Pull request merged**